### PR TITLE
implement filter3D

### PR DESCRIPTION
**Pull request by edgarriba**
### Description
This PR implements filter3D which is an extension for filter2d to handle filtering in volumetric data.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Comment by wyli** [comment]
thanks, we might want to have interfaces for separable filters as well (https://en.wikipedia.org/wiki/Separable_filter)?  the current implementation is generic but will be slow for relatively large kernels

**Comment by ducha-aiki** [comment]
@wyli as far as I have tested, as least for 2d, separable filters were slow, probably because of lack of optimization in cudnn

File: kornia/filters/filter.py [hunk 1]
```diff
@@ -71,16 +87,119 @@ def filter2D(input: torch.Tensor, kernel: torch.Tensor,
...
+
+    if normalized:
+        bk, dk, hk, wk = kernel.shape
+        tmp_kernel = normalize_kernel2d(tmp_kernel.view(
+            bk, dk, hk * wk)).view_as(kernel)
+
+    # pad the input tensor
+    depth, height, width = tmp_kernel.shape[-3:]
+    padding_shape: List[int] = compute_padding([depth, height, width])
+    input_pad: torch.Tensor = F.pad(input, padding_shape, mode=border_type)
```
Comment (ducha-aiki) [inline kornia/filters/filter.py:202]: We may want to get rid of explicit padding when all Conv*d functions would support all types of padding - that would save a lot of memory.

**Comment by wyli** [comment]
> @wyli as far as I have tested, as least for 2d, separable filters were slow, probably because of lack of optimization in cudnn

is that the case for 3D? I think this PR is about adding the 3D support

**Comment by ducha-aiki** [comment]
Yes. What I mean that we should benchmark first

**Comment by edgarriba** [comment]
@wyli besides, how would you shape the API for  separable convs? I guess it would work only for "symmetric" 1d kernels.

**Pull request merged**